### PR TITLE
Set etag to be null if it's empty

### LINKED ISSUE #405: States without etag is handled differently in gRPC and HTTP APIs
## Expected Behavior

Get a state without eTag in gRPC and HTTP, eTag attribute should be null.

## Actual Behavior

eTag is null in HTTP and empty string in gRPC.

## Steps to Reproduce the Problem

Get state that does not have eTag.
Try the same in getBulkState too.

## Release Note

RELEASE NOTE: **FIX** missing eTag response for state APIs.

### PR DESCRIPTION
**Pull request by arghya88**
Signed-off-by: Arghya Sadhu <arghya88@gmail.com>

# Description

Convert empty string to null for GRPC implementation to keep it consistent with HTTP implementation

## Issue reference

Fixes #405

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

**Review by mukundansundar — COMMENTED** [review]
lgtm.
@artursouza The issue linked to this should be a breaking change since the response is changing.

**Comment by artursouza** [comment]
Yes, LGTM. Let's mark it as breaking change - even though it is minor.

File: sdk/src/main/java/io/dapr/client/DaprClientGrpc.java [hunk 1]
```diff
@@ -332,6 +332,9 @@
     byte[] data = payload == null ? null : payload.toByteArray();
     T value = stateSerializer.deserialize(data, type);
     String etag = response.getEtag();
+    if (etag.equals("")) {
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DaprClientGrpc.java:335]: Bulk state needs this change too. See method above this one.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/dapr/client/DaprClientGrpc.java (hunk 1 above)
Comment (mukundansundar) [reply sdk/src/main/java/io/dapr/client/DaprClientGrpc.java:335]: good catch. missed that ..

Comment (arghya88) [reply sdk/src/main/java/io/dapr/client/DaprClientGrpc.java:338]: Done

File: sdk-tests/src/test/java/io/dapr/it/state/AbstractStateClientIT.java [hunk 2]
```diff
@@ -109,12 +109,12 @@ public void saveAndGetBulkStates() {
 
     assertEquals(stateKeyTwo, result.stream().skip(1).findFirst().get().getKey());
     assertNull(result.stream().skip(1).findFirst().get().getValue());
-    assertEquals("1", result.stream().skip(1).findFirst().get().getEtag());
+    assertNull(result.stream().skip(1).findFirst().get().getEtag());
```
Comment (arghya88) [inline sdk-tests/src/test/java/io/dapr/it/state/AbstractStateClientIT.java:112]: @artursouza I wonder how this assert was successful before.

Comment (artursouza) [reply sdk-tests/src/test/java/io/dapr/it/state/AbstractStateClientIT.java]: Ops. Let me see if this is a bug that was fixed or created.

Comment (artursouza) [reply sdk-tests/src/test/java/io/dapr/it/state/AbstractStateClientIT.java]: I see that his change is not in anymore. So I assume it was a red herring.

**Comment by arghya88** [comment]
@artursouza I debugged why `saveAndGetBulkState` test in `AbstractStateClientIT` is failing. It turns out

1. For the second state object both GRPC and HTTP client returns `1` for Etag even though we set `null` in the object. 
2. For the third object where we don't even save it in state store HTTP client returns empty and GRPC client returns `null` for Etag.

I am not sure if 1 is expected to be this way. If this is expected then we can keep it this way and let the test pass.
For 2 the tests used to pass before my change because GRPC client used to return empty before my change. With my change it returns null instead of empty and hence  test for GRPC client is failing.

**Comment by arghya88** [comment]
@artursouza Pls review

**Comment by artursouza** [comment]
> @artursouza I debugged why `saveAndGetBulkState` test in `AbstractStateClientIT` is failing. It turns out
> 
> 1. For the second state object both GRPC and HTTP client returns `1` for Etag even though we set `null` in the object.
> 2. For the third object where we don't even save it in state store HTTP client returns empty and GRPC client returns `null` for Etag.
> 
> I am not sure if 1 is expected to be this way. If this is expected then we can keep it this way and let the test pass.
> For 2 the tests used to pass before my change because GRPC client used to return empty before my change. With my change it returns null instead of empty and hence test for GRPC client is failing.
> 
> 2 is not happening in unit tests because units tests don't cover a scenario of trying to retrieve a state object which is not saved at all.

The reason why the etag is "1" and not null even when we set it to null is because Dapr maps etag to version in Redis. Version is always present in Redis, so it defaults to "1" when we set a value: https://github.com/dapr/components-contrib/blob/master/state/redis/redis.go

This is why the integration tests asserts "1" for the null case.

**Review by artursouza — APPROVED** [review]

**Pull request merged**